File: merkle-trie/src/main/java/net/consensys/cava/trie/LeafNode.java
package net.consensys.cava.trie;

import static net.consensys.cava.crypto.Hash.keccak256;

import net.consensys.cava.bytes.Bytes;
import net.consensys.cava.bytes.Bytes32;
import net.consensys.cava.concurrent.AsyncResult;
import net.consensys.cava.rlp.RLP;

import java.lang.ref.SoftReference;
import java.lang.ref.WeakReference;
import java.util.Optional;
import java.util.function.Function;

final class LeafNode<V> implements Node<V> {
  private final Bytes path;
  private final V value;
  private final NodeFactory<V> nodeFactory;
  private final Function<V, Bytes> valueSerializer;
  private WeakReference<Bytes> rlp;
  private SoftReference<Bytes32> hash;

  LeafNode(Bytes path, V value, NodeFactory<V> nodeFactory, Function<V, Bytes> valueSerializer) {
    this.path = path;
    this.value = value;
    this.nodeFactory = nodeFactory;
    this.valueSerializer = valueSerializer;
  }

  @Override
  public AsyncResult<Node<V>> accept(NodeVisitor<V> visitor, Bytes path) {
    return visitor.visit(this, path);
  }

  @Override
  public AsyncResult<Bytes> path() {
    return AsyncResult.completed(path);
  }

  @Override
  public AsyncResult<Optional<V>> value() {
    return AsyncResult.completed(Optional.of(value));
  }

  @Override
  public Bytes rlp() {
    if (rlp != null) {
      Bytes encoded = rlp.get();
      if (encoded != null) {
        return encoded;
      }
    }

    Bytes encoded = RLP.encodeList(writer -> {
      writer.writeValue(CompactEncoding.encode(path));
      writer.writeValue(valueSerializer.apply(value));
    });
    rlp = new WeakReference<>(encoded);
    return encoded;
  }

  @Override
  public Bytes rlpRef() {
    Bytes rlp = rlp();
    if (rlp.size() < 32) {
      return rlp;
    } else {
      return RLP.encodeValue(hash());
    }
  }

  @Override
  public Bytes32 hash() {
    if (hash != null) {
      Bytes32 hashed = hash.get();
      if (hashed != null) {
        return hashed;
      }
    }
    Bytes32 hashed = keccak256(rlp());
    hash = new SoftReference<>(hashed);
    return hashed;
  }

  @Override
  public AsyncResult<Node<V>> replacePath(Bytes path) {
    return nodeFactory.createLeaf(path, value);
  }
}

File: merkle-trie/src/main/kotlin/net/consensys/cava/trie/experimental/LeafNode.kt
/*
 * Copyright 2018, ConsenSys Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.consensys.cava.trie.experimental

import net.consensys.cava.bytes.Bytes
import net.consensys.cava.bytes.Bytes32
import net.consensys.cava.crypto.Hash.keccak256
import net.consensys.cava.rlp.RLP
import net.consensys.cava.trie.CompactEncoding
import java.lang.ref.SoftReference
import java.lang.ref.WeakReference

internal class LeafNode<V>(
  private val path: Bytes,
  private val value: V,
  private val nodeFactory: NodeFactory<V>,
  private val valueSerializer: (V) -> Bytes
) : Node<V> {
  private var rlp: WeakReference<Bytes>? = null
  private var hash: SoftReference<Bytes32>? = null

  override suspend fun accept(visitor: NodeVisitor<V>, path: Bytes): Node<V> = visitor.visit(this, path)

  override suspend fun path(): Bytes = path

  override suspend fun value(): V? = value

  override fun rlp(): Bytes {
    val prevEncoded = rlp?.get()
    if (prevEncoded != null) {
      return prevEncoded
    }

    val encoded = RLP.encodeList { writer ->
      writer.writeValue(CompactEncoding.encode(path))
      writer.writeValue(valueSerializer(value))
    }
    rlp = WeakReference(encoded)
    return encoded
  }

  override fun rlpRef(): Bytes {
    val rlp = rlp()
    return if (rlp.size() < 32) rlp else RLP.encodeValue(hash())
  }

  override fun hash(): Bytes32 {
    val prevHashed = hash?.get()
    if (prevHashed != null) {
      return prevHashed
    }
    val hashed = keccak256(rlp())
    hash = SoftReference(hashed)
    return hashed
  }

  override suspend fun replacePath(path: Bytes): Node<V> = nodeFactory.createLeaf(path, value)
}
